File: benchmarks/tsb/bench_interpolate_methods.ts
/**
 * Benchmark: interpolateSeries with linear, ffill, bfill, nearest, zero methods.
 * Outputs JSON: {"function": "interpolate_methods", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Series, interpolateSeries } from "../../src/index.ts";

const SIZE = 50_000;
const WARMUP = 5;
const ITERATIONS = 30;

// Build a series with ~20% NaN scattered
const data: (number | null)[] = Array.from({ length: SIZE }, (_, i) =>
  i % 5 === 0 ? null : i * 0.1,
);
const s = new Series({ data });

for (let i = 0; i < WARMUP; i++) {
  interpolateSeries(s, { method: "linear" });
  interpolateSeries(s, { method: "ffill" });
  interpolateSeries(s, { method: "bfill" });
  interpolateSeries(s, { method: "nearest" });
  interpolateSeries(s, { method: "zero" });
}

const times: number[] = [];
for (let i = 0; i < ITERATIONS; i++) {
  const start = performance.now();
  interpolateSeries(s, { method: "linear" });
  interpolateSeries(s, { method: "ffill" });
  interpolateSeries(s, { method: "bfill" });
  interpolateSeries(s, { method: "nearest" });
  interpolateSeries(s, { method: "zero" });
  times.push(performance.now() - start);
}

const totalMs = times.reduce((a, b) => a + b, 0);
const meanMs = totalMs / ITERATIONS;
console.log(
  JSON.stringify({
    function: "interpolate_methods",
    mean_ms: Math.round(meanMs * 1000) / 1000,
    iterations: ITERATIONS,
    total_ms: Math.round(totalMs * 1000) / 1000,
  }),
);

File: benchmarks/pandas/bench_interpolate_methods.py
"""Benchmark: interpolateSeries with linear, ffill, bfill, nearest, zero methods."""
import json, time
import pandas as pd
import numpy as np

SIZE = 50_000
WARMUP = 5
ITERATIONS = 30

data = [float(i) * 0.1 if i % 5 != 0 else None for i in range(SIZE)]
s = pd.Series(data, dtype=float)

for _ in range(WARMUP):
    s.interpolate(method="linear")
    s.ffill()
    s.bfill()
    s.interpolate(method="nearest")
    s.interpolate(method="zero")

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    s.interpolate(method="linear")
    s.ffill()
    s.bfill()
    s.interpolate(method="nearest")
    s.interpolate(method="zero")
    times.append((time.perf_counter() - t0) * 1000)

total_ms = sum(times)
print(json.dumps({"function": "interpolate_methods", "mean_ms": round(total_ms / ITERATIONS, 3), "iterations": ITERATIONS, "total_ms": round(total_ms, 3)}))
